bindings/rascal/utils/io.py

```diff
@@ -0,0 +1,376 @@
+    """
+    cls = get_class(data['module_name'], data['class_name'])
+    obj = cls(**data['init_params'])
+    obj._set_data(data['data'])
+    return obj
+
+
+def is_valid_object_dict_beta(data):
+    """check compatibility of data to be used in dict2obj_beta"""
+    valid_keys = ['version', 'class_name', 'module_name', 'init_params', 'data']
+    aa = []
+    if isinstance(data, dict):
+        for k in data:
+            if k in valid_keys:
+                aa.append(True)
+        if len(aa) == len(valid_keys):
+            return True
+        else:
+            return False
+    else:
+        return False
+
+
+obj2dict = {BETA_VERSION: obj2dict_beta}
+dict2obj = {BETA_VERSION: dict2obj_beta}
+is_valid_object_dict = {BETA_VERSION: is_valid_object_dict_beta}
+
+
+def get_current_io_version():
+    return CURRENT_VERSION
+
+
+def get_supported_io_versions():
+    return list(dict2obj.keys())
+
+
+class BaseIO(ABC):
+    """
+    Definition of the interface of a python class that can be serialized by the
+    to_dict() function. It corresponds to 3 methods:
+
+    + _get_init_params is expected to return a dictionary containing all the
+    parameters used by the __init__() methods.
+
+    + _get_data is expected to return a dictionary containing all the data
+    that is not set by the initialization of the class.
+
+    + _set_data is expected to set the data that has been extracted by _get_data
+    """
+    @abstractmethod
+    def _get_data(self):
+        return dict()
+
+    @abstractmethod
+    def _set_data(self, data):
+        pass
+
+    @abstractmethod
+    def _get_init_params(self):
+        return dict()
+
+
+def _get_state(obj):
+    if isinstance(obj, BaseIO) is True:
+        state = dict(data=obj._get_data(),
+                     init_params=obj._get_init_params())
+    else:
+        raise ValueError(
+            'input object: "{}" does not inherit from "BaseIO"'.format(obj))
+    return state
+
+
+def to_dict(obj, version=CURRENT_VERSION):
+    """recursirvely convert the python object obj into a dictionary (serialized
+    version).
+    obj has to inherit from BaseIO."""
+
+    state = _get_state(obj)
+
+    # loop over the 2 fields of state
+    for name, entry in state.items():
+        if isinstance(entry, dict):
+            # case of potentially nested objects
+            for k, v in entry.items():
+                if isinstance(v, BaseIO) is True:
+                    state[name][k] = to_dict(v, version)
+                elif isinstance(v, list):
+                    # make sure list of objects are properly serialized
+                    ll = []
+                    for val in v:
+                        if isinstance(val, BaseIO) is True:
+                            ll.append(to_dict(val, version))
+                        else:
+                            ll.append(val)
+                    state[name][k] = ll
+    data = obj2dict[version](obj.__class__, state)
```

Comment:
Does that work for 'weird' python objects too? Like 

```python
# circular object
foo = {"bar: 3}
foo["blee"] = foo

# object containing C ressources (pointers, file descriptor, ...)
bar = {"file": open('tmp.xyz')}
```

Reply:
> # circular object

I don't think we have a reasonable use case for suporting this added complexity so not at the moment. It is not hard to suport but again I would rather not implement it. Are you thinking of making a check so that en error is raised ?

> # object containing C ressources (pointers, file descriptor, ...)

file descriptor would not be possible. But the filename could be handled like it is done for large numpy arrays. 
As for objects that contain C++ resources, they have to call a `to_dict()` function implemented in the bindings that should use the c++ json library to achieve that. What use case do you have in mind exactly ?

Reply:
I don't have specific use-cases right now, these are mainly tricky things to get right when doing Python serialization. So having a check for them would be good, or even better explicitly disallowing everything we don't know about. I would only allow string, numbers, list of the previous and dict of the previous (taking care of circular objects); and throw an error on anything else.

Reply:
Do you have an idea on how to detect circularity ? setting a max depth of serialization seems simple but is there a better way ?

Reply:
not on the top of my mind. I know the json module does it, so you could take inspiration from there (or just piggy-back on it `try: json.dump(x); except ValueError as e: if str(e) == "Circular reference detected" raise`)

Reply:
ok I will set a max recursion parameter and raise for circular reference when it is reached